### ISSUE TITLE
rust-analyzer.toml

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#13529]
This is a tracking issue/discussion for a potential `rust-analyzer.toml`. There have been a lot of requests for a per-project configuration in various issues, some clients like VSCode allow specifying settings per workspace, but usually these kinds of settings are not commited in a repo and are therefor usually not meant to necessarily configure r-a for the project specifically (like certain coding style settings, think of import merging etc.). Other related things might be requiring a special built procedure for a cargo project, prime example being https://github.com/rust-lang/rust which currently has a few suggested config lines in the dev guide which a user has to specify manually when they check out the repo locally, see https://rustc-dev-guide.rust-lang.org/building/suggested.html#configuring-rust-analyzer-for-rustc.

So there is clearly a need for a project config option like this, but here are few open questions.

- [x] We could make use of the `Cargo.toml` metadata section instead, but that would only solve the problem for cargo based projects and it feels wrong to have certain settings in there opposed to a `rust-analyzer.toml` I believe.
  - I believe a `rust-analyzer.toml` to be the correct solution here. 
- [ ] How does the priority work in regards to other mechanisms that set configurations? A `rust-analyzer.toml` should probably overwrite the client's configuration, as we cannot know if a client sents us the default config for something or if its explicitly set.
- [ ] How would r-a look for these, is nesting allowed, that is can we have hierarchies for these files? We should probably stick to what other tools like rustfmt do in this regard. Hierarchies could pose an implementation problem though, as r-a only has global configurations (global in the sense of project wide, not per workspace). I am unsure how complex it would be to change that, so this might entail a bigger architecture change.

**Comment by Qix-** [comment]
Note: Highly suggest making this a dotfile (`.rust-analyzer.toml`).

**Comment by emilio** [comment]
Yes please.

We need something like this to provide proper integration with our build system and right now we can only provide that for [vscode](https://searchfox.org/mozilla-central/rev/26790fecfcda622dab234b28859da721b80f3a35/python/mozbuild/mozbuild/backend/mach_commands.py#321-328) which is unfortunate.

>  How does the priority work in regards to other mechanisms that set configurations? Ideally we'd favor a client's settings over rust-analyzer.toml, but we cannot know if a client sends us the default config for something or if its explicitly set, so we might have to bite the bullet and prefer rust-analyzer.toml.

FWIW clangd has this feature ([`.clangd` files](https://clangd.llvm.org/config#files)) so may be worth checking out.

**Comment by Qix-** [comment]
> How does the priority work in regards to other mechanisms that set configurations?

Doesn't this depend on how the client sends configs at all? Does r-a get its configuration via the client, or from the filesystem? If the former, then it's up to the client implementations to specify this. Ideally each would allow the user to specify the priorities of the searched configs, be able to merge them based on this, and then send the flattened config to r-a. I don't see how that's r-a's problem, really.

> How would r-a look for these, is nesting allowed, that is can we have hierarchies for these files?

I don't think nesting is necessary. Keeping it clean and simple is probably the best route. It would also make debugging easier. (Again, this assumes that r-a is getting the config from the client directly, not from scouring the source tree etc.)

> What configs should be exposed in this

If the above are true, then is this really a fix for r-a directly? It'd be up to the individual client implementations to handle these sort of semantics. Perhaps r-a could specify that `.rust-analyzer.toml` is a valid configuration spot and specify its syntax. Otherwise, the client libs should probably have the "sane" default of...

1. Any in-client overrides (directly specified in the config of the client)
2. Any user-local files (e.g. `~/.rust-analyzer.toml`), if it exists
3. `$(pwd)/.rust-analyzer.toml`, if it exists
4. Some default set of configuration pertinent to getting r-a working in the respective client

As far as I understand, 1. and 4. are already the case. It's 2. and 3. that would have to be added.

Bonus points for allowing the user to specify search paths / their order for `.rust-analyzer.toml` files, allowing for both absolute and relative paths.

**Comment by Veykril** [comment]
My initial thought here was for the server to consume the `rust-analyzer.toml`, not the client. Having the client consume it seems pointless, as that requires any LSP client to add special support for the `rust-analyzer.toml` format and ideally, rust-analyzer should have a baseline usability with any LSP compliant client.

---


> Doesn't this depend on how the client sends configs at all? Does r-a get its configuration via the client, or from the filesystem? If the former, then it's up to the client implementations to specify this. Ideally each would allow the user to specify the priorities of the searched configs, be able to merge them based on this, and then send the flattened config to r-a. I don't see how that's r-a's problem, really.

So effectively with this `rust-analyzer.toml` added, we would have two ways of receiving config data, by the client and then per-project (per cargo workspace, opened project folder or crate even etc). I think we should always prefer the `rust-analyzer.toml` settings in that regard, the main reason I raised this question was because VSCode for example has global settings and local project settings but we can only treat them as "the client settings" in one unit, but that should be fine I think. Also note that not all config settings exposed by r-a today would make sense in the `rust-analyzer.toml`, an example being `cargo.autoreload`.

> Bonus points for allowing the user to specify search paths / their order for .rust-analyzer.toml files, allowing for both absolute and relative paths.

That's a good point to consider.

---

The other points you've raised are somewhat moot with the thought of the server consuming it (I shouldve been clearer on that in the issue description, will edit) I think.

**Comment by Veykril** [comment]
> Note: Highly suggest making this a dotfile (`.rust-analyzer.toml`). This would be aligned with other project-specific modifier configs such as `.rustfmt.toml` and `.rusty-hook.toml`, etc.

Allowing both `rust-analyzer.toml` and `.rust-analyzer.toml` seems fine to me (and given `rustfmt` consumes either forms as well it would only make sense for us to do the same)

**Comment by swarnimarun** [comment]
I remember another request for allowing Rust Analyzer to work without `Cargo.toml` aka over a 
single file, I believe some config options were still required could we add them to `rust-analyzer.toml` would make scripting very easy if we could also have a `rust-analyzer.toml` globally. For say default custom global configuration?

And ensure that we use `Cargo.toml` or folder-local config only if they are present?

**Comment by Veykril** [comment]
Yes that somewhat plays into the third checkbox point, how r-a will look for the file and if nesting is allowed.

**Comment by alibektas** [comment]
Here is more or less a roadmap that I plan to follow : (The two names that I use here `global` and `local` are not very descriptive and probably will be changed to something more meaningful so do not mind them for now. And let `.rust-analyzer.toml` be `RATOML` for now on. )

- [ ] Find out which configs are `global` or `local`-scoped. Create a PR where we possibly publish a document in which we briefly mention why a certain key-value pair is `global` or `local`.
- [ ] Replace `rust_analyzer::config::ConfigData` with `GlobalConfigData` `LocalConfigData`. As a result of this change `rust_analyzer::config::Config`'s `data` will be of type `GlobalConfigData`. `project_model::ProjectManifest` is a good place to handle the discovery of the `.rust-analyzer.toml` file and finding out which configurations apply to a single `ProjectManifest`. This change will probably entail adding to more variants to `ProjectManifest` like `ProjectManifest::ProjectJsonWithConf` and `ProjectManifest::CargoTomlWithConf`. As I said before names are at this point arbitrary.
- [ ] Client settings and `RATOML` : During first handshakes between the server and the client config data are sent from the client to the server, which is how we have received configuration data so far. I plan to keep this as is and let server do a fresh start if any `RATOML` files are discovered. But this shouldn't tell you that the configs sent from the client have a higher priority than `RATOML`. Although I agree with @Veykril on the point that our hand is forced to choose either `RATOML` or client settings over another, the user should still be able to say that they want to use their own configs, which means that we need to add a KV pair that does just that : telling us which source should we prefer.
- [ ] First version of the `RATOML` : Although I may have sounded like this is something I want to have ready in the first version, I think we should first enable the use of a single `RATOML` file and gradually introduce nesting. So this version will use two sources to read configs from : (1) `lsp_types::InitializeParams` as sent by the client (2) `RATOML`.

**Comment by Qix-** [comment]
What does `global` and `local` solve? Why not just use a list of configs that overwrite each other in order, as I mentioned before?

**Comment by alibektas** [comment]
There are some configurations for which overwriting doesn't make much sense and can even be misleading. An example would be `rust-analyzer.cachePriming.enable` which states whether to `Warm up caches on project load.` ( this example may also be the wrong example I am not super sure about it but the point I am trying to make is that some configs are different from other in that they should be configured once). As to what goes in the `RATOML` there is no difference between globals and locals, they are configured exactly the same way it is just that for a global **you can still nest it but only its first occurrence counts.**

**Comment by Qix-** [comment]
I still don't see why global vs local makes more sense than multiple configs. Who cares if a second config overwrites the first in some weird way? That's on the user at that point.

**Comment by Veykril** [comment]
(That distinction came from a discussion Ali and I had elsewhere)

It is more that there a configs that apply to the server session as a whole and then there are configs that apply to individually loaded projects. There are also configs that do not make sense to be in the rust-analyzer.toml at all, like all the hover configs for example. Those are project irrelevant so they won't be configurable by the r-a toml file.

Though having a second look at things, I don't think there are configs that can't be applied on a per loaded project level (opposed to server session wide), aside from current implementation reasons maybe (`files.watcher` would be a server wide one, but at the same time that one should not be configured by the `rust-analyzer.toml`). So ye, the first step here is to classify all the current configs into the three buckets `not applicable for rust-analyzer.toml`, `project/cargo workspace applicable` and `"global"/session wide`. Then we can see whether that split makes sense or not.

**Comment by Veykril** [comment]
https://github.com/rust-lang/rust-analyzer/pull/17058 will implement the core features here, there are still some issues with it (the user config is not working correctly yet), as well some restrictions with crate specific configs due to architecture reasons (they shouldn't really be noticeably in most cases). A lot of configs are also still missing from it and the diagnostics UX (when an invalid config is supplied) is really bad. All of that will be resolved in the near or far future (time will tell).

> [!IMPORTANT]  
> We don't promise ***any*** stability with this feature yet, any configs exposed may be removed again, the ordering may change etc.


Once the PR lands (should today), this issue will be used to track any updates on the feature.

**Comment by kornelski** [comment]
A rust-analyzer config file, or at least something in `Cargo.toml` would be great. I often need to enable most but not all non-default features of projects I'm working on to get useful autocomplete. 

BTW: the rust-analyzer project is incredibly biased towards VSCode-only solutions. As someone who deliberately doesn't use this editor, I'm constantly frustrated by features that are missing from rust-analyzer, because they're solved by some VSCode-only doodad, and are unsupported or undocumented for other editors. Per-project config is one of these.

**Comment by alibektas** [comment]
@kornelski :wave: you can already use per-project configuration ( see #17058 ), there are two known issues to this which will be soon resolved ( see #17483 ). Please let me know if you have any suggestions/comments in general.

**Comment by Veykril** [comment]
> BTW: the rust-analyzer project is incredibly biased towards VSCode-only solutions.

It is and we can't do anything about it. We are limited by what the LSP allows us to do. VSCode is the reference implementation that we maintain, we can't possible maintain editor extensions for all editors out there for obvious reasons.

**Comment by kornelski** [comment]
@alibektas that's a great news. However, I can't figure out what syntax it wants. I've tried `rust-analyzer.toml` in the project root with:

```toml
[rust-analyzer.cargo]
features = "all"
# or
[cargo]
features = "all"
```

but I don't see any effect, even after restarting RA manually.

@Veykril The problem is that if you don't make RA less VSCode-centric, nobody else will do that for you. The other editors say "this is just one of many LSP clients, we can't possibly test and document all LSP implementations", and RA outside of VSCode ends up in a no man's land of undocumented and untested integrations owned by nobody.

**Comment by flodiebold** [comment]
@kornelski As a non-VSCode user, I don't agree that RA is VSCode-centric. Yes, the reference implementation is the VSCode plugin, but, well, something needs to be. The RA team can't possibly maintain integrations for a bunch of editors no one on the team uses. If the integration for a specific editor is bad, someone from that community needs to step up to maintain it.

> The other editors say "this is just one of many LSP clients, we can't possibly test and document all LSP implementations",

Maybe don't use those editors then if you want good LSP integration. My experience is that LSP has not been designed with the goal of "no per-server integration code needed in the editor at all".

> I'm constantly frustrated by features that are missing from rust-analyzer

Can you give some examples of those features? (Except per-project config, which is literally already implemented)

**Comment by matklad** [comment]
FWIW, i'd love to see `./editors/helix`, `./editors/emacs`, `./editors/nvim` in this repository. That's why it was, forever, `./editors/code` and not just `./code`. But that of course requires:

* someone on the team to maintain those extensions
* editor-side infrastructure to allow extensions. 

>My experience is that LSP has not been designed with the goal of "no per-server integration code needed in the editor at all".

Yes. This is absolutely the design. This is a very important fact: there is _no_ LSP support directly in VS Code. There are only language-specific extensions, which happen to use the same common LSP library (that is, each extension bundles its own private copy of the library).

---

All our extras are documented here:

* https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md
* there's "RSS feed" for updates here: https://github.com/rust-lang/rust-analyzer/issues/4604

**Comment by rzvxa** [comment]
> @alibektas that's a great news. However, I can't figure out what syntax it wants. I've tried `rust-analyzer.toml` in the project root with:
> 
> ```toml
> [rust-analyzer.cargo]
> features = "all"
> # or
> [cargo]
> features = "all"
> ```
> 
> but I don't see any effect, even after restarting RA manually.

@alibektas What is the intended syntax I also have issues with setting it up

In addition to the 2 ways mentioned above, I've also tried this to no avail.

```toml
cargo.features = all
```

I get this error.

```
invalid config value:
cargo/features: unexpected field
```

**Comment by lnicola** [comment]
@alibektas is something like `cachePriming.enable = false` supposed to work? I think `assist.emitMustUse = true` is the only one used in tests, and it works for me, but not the others. See also rzvxa's comment above.

**Comment by lnicola** [comment]
`~/.config/rust-analyzer/rust-analyzer.toml` doesn't seem to get loaded in my tests.

**Comment by alibektas** [comment]
> `~/.config/rust-analyzer/rust-analyzer.toml` doesn't seem to get loaded in my tests (options don't take effect, and it doesn't complain about syntax errors).

That I still need to fix. I worked on it but it is still not complete

**Comment by alibektas** [comment]
You were right in assuming that we omit `rust-analyzer` prefix. I will add an example to #17058. A thorough documentation will follow as soon as we update our manual. The reason why it does not have an affect may simply be that it is a bug. I will take a look at it today and will let you know. Ok?

**Comment by Logarithmus** [comment]
Where can I find an example of `rust-analyzer.toml`? Can't find any documentation whatsoever

**Comment by alibektas** [comment]
> Where can I find an example of `rust-analyzer.toml`? Can't find any documentation whatsoever

We still don't have a dedicated documentation to it. I guess you can see #17058 or ask me if you have any questions.

**Comment by bluss** [comment]
Would it be possible for you to show an example of a valid, working `rust-analyzer.toml` file in a comment here? So that syntax and structure becomes clear. An example would go a long way.

**Comment by alibektas** [comment]
The first to keep in mind that `rust-analyzer.toml` is still in progress. You may not get the intended behavior just yet. Anything where you get something unexpected deserves a separate issue :) 

`rust-analyzer.toml` ( a.k.a `ratoml` ) is a a file that follows the `toml` syntax. Its keys are those specified in the `rust-analyzer` [manual](https://rust-analyzer.github.io/manual.html), only difference being that the `rust-analyzer.` prefix is always omitted. 

So sth like 

```json
{
      "rust_analyzer.cachePriming.enable" = true,
      "rust_analyzer.files.excludeDirs" = [ "abc" ], 
}
``` 

becomes 

```toml
cachePriming.enable = true
files.excludeDirs = [ "abc" ]
```

One advantage of toml's syntax over that of `json`'s is that we can group keys with mutual prefixes in a table. So sth like

```json
{
        "rust_analyzer.highlightRelated.breakPoints.enable"  = true,
        "rust_analyzer.highlightRelated.closureCaptures.enable"  =  true,
        "rust_analyzer.highlightRelated.exitPoints.enable" = true,
        "rust_analyzer.highlightRelated.references.enable" = true,
        "rust_analyzer.highlightRelated.yieldPoints.enable"  = true,
}
```

can easily be turned into 

```toml
[highlightRelated]
breakPoints.enable = true
closureCaptures.enable = true
exitPoints.enable = true
references.enable = true
yieldPoints.enable = true
```

It is a trivial thing but I still wanted to mention it :) 
At this point I want to say two things :  There must be many problems with the current `ratoml` infrastructure. It is getting better everyday but we are still not in a comfortable state with it. 

Since `ratoml` offers a much more flexible configurability as opposed to editor config files such as `settings.json` in VSCode where you have a single file to configure an entire project, it was necessary to adapt rest of the codebase to this new condition. It is important to check once again the issue #17058 to see what kind of `config classes` we thought of. Each config key will belong to one of these classes and this property is important because it defines the domain in which a key-value pair takes affect. So if for instance a key is a `workspace` key, then it will only take affect in the workspace it is defined, and it can only be defined at the root of a workspace.

**Comment by csmulhern** [comment]
Using `rust-analyzer 0.0.0 (a341d598d 2024-11-04)`, I'm unable to get a `rust-analyzer.toml` file to be recognized at the root of the workspace.

If I run:
```
cargo new example
cd example
echo "assist.emitMustUse = true" > rust-analyzer.toml
```

And then start my editor (Zed) with `env RA_LOG=rust_analyzer=info`, I see that no ratoml file is detected.

```
// logs
...
ratoml_file: {},
...
```

Are we missing resolution of ratoml files at the root of the workspace?

**Comment by alibektas** [comment]
> Using `rust-analyzer 0.0.0 (a341d598d 2024-11-04)`, I'm unable to get a `rust-analyzer.toml` file to be recognized at the root of the workspace.
> 
> If I run:
> ```
> cargo new example
> cd example
> echo "assist.emitMustUse = true" > rust-analyzer.toml
> ```
> 
> And then start my editor (Zed) with `env RA_LOG=rust_analyzer=info`, I see that no ratoml file is detected.
> 
> ```
> // logs
> ...
> ratoml_file: {},
> ...
> ```
> 
> Are we missing resolution of ratoml files at the root of the workspace?

👋 Is there a Cargo.toml file in the same directory?

**Comment by alibektas** [comment]
Ok just deleted my previous comment I will check and see what's wrong. This kind of an error is ofc not supposed to have happened

**Comment by alibektas** [comment]
@csmulhern  Ok so I checked it. Every is working fine (at least for this particular config key) . The reason why file names don't show up is that they are not detected at the stage where you get this debug output. Let me know if you have any other questions/simply don't agree with this :)

**Comment by alibektas** [comment]
@csmulhern Oh I see. You don't have to open a separate issue for this config key because it is already known to me and I am working on this, but there is a preliminary work to be done (see 18186 I don't want to create a link to this issue so pardon me for the inconvenience). It should take about some three weeks as the team-lead is away at the moment and he is the only other person who could actually review my PR ( PR is also WIP so you won't find anything yet). Until then I can sadly recommend you to use your editor's config file to configure rust-analyzer. I will keep you posted, ok?

**Comment by dustypomerleau** [comment]
> My concern is that TOML does not support null.
> 
> > rust-analyzer.inlayHints.maxLength (default: 25)
> > Maximum length for inlay hints. Set to null to have an unlimited length.
> >
> > https://rust-analyzer.github.io/manual.html#rust-analyzer.inlayHints.maxLength
> 
> How can I set `rust-analyzer.inlayHints.maxLength` to null?
> 

Have you tried setting the value to `"null"`? It looks like the RA config macros stringify the fields prior to parsing them.

**Comment by mamekoro** [comment]
I tried, but it didn't work.

There should be no ellipsis (`…`) if `maxLength` is successfully set to unlimited.
![Image](https://github.com/user-attachments/assets/b485bdce-74d0-4dc2-9a7a-6d572bd56168)

For your reference, below are the screenshots when `maxLength` is 0 and 50.
![Image](https://github.com/user-attachments/assets/8e932db7-6f18-4c75-85a6-7c0c0333ed10)
![Image](https://github.com/user-attachments/assets/7f0bc650-9738-48da-8717-433bd58667b4)

**Comment by codeitlikemiley** [comment]
actually there are two problems here, one is ra for server and the other one is client...

I have built a vscode plugin called [cargo-runner](https://github.com/codeitlikemiley/cargo-runner)

 that solve the issue on client side

issue for running commands you manually triggered 

like run, test, bench , debug can be handle and override per crate using my .cargo-runner.toml file

while server side involves build,  and check

which usually has issue or not respecting per crate rust-toolchain.toml or in this case per project config of ratoml

is the problem , that i usually add extra hack on my plugin to add necessary metadata when runnign cargo check on per crate 
but that usually causes the server to restart , not good developer experience as you need to restart everything again you switch those ratoml server related config.

My goal of the plugin im developing is become a partner for ra, 

but some might be added to ra , 

but atm , i need to study the ra repo , 

if i wanna create those patches i had in mind...

**Comment by Veykril** [comment]
Headsup on the status of this, the feature is currently pretty broken (and https://github.com/rust-lang/rust-analyzer/pull/18949 will likely make it work less), the reasons for that are architectural which unfortunately means we need to rewrite a decent part (https://github.com/rust-lang/rust-analyzer/issues/18948) of code.

**Comment by dlight** [comment]
Is it possible to make a way to configure it per-workspace, for projects with multiple workspaces? Something like a `rust-analyzer.toml` at the root of each workspace.

**Comment by lucasbaizer2** [comment]
> You were right in assuming that we omit `rust-analyzer` prefix. I will add an example to #17058. A thorough documentation will follow as soon as we update our manual. The reason why it does not have an affect may simply be that it is a bug. I will take a look at it today and will let you know. Ok? 
> 
> EDIT : Now thinking about it for a second time, the problem must be due to the fact that `cargo.features` is defined as `workspace` meaning that it has no affect when it is defined on a crate level. So we need to make it local I guess. I will think about this as I said.

Is there any update on configuring per-crate features? Maybe another GitHub issue we can track?
